### PR TITLE
[manila]: implement share transfer API

### LINKED ISSUE #2647: [manila]: implement share transfer API
https://docs.openstack.org/api-ref/shared-file-system/?expanded=#share-transfer-since-api-v2-77

Mostly it's the same as transferring the Cinder Volumes, but for shares: https://github.com/gophercloud/gophercloud/blob/master/openstack/blockstorage/extensions/volumetransfers/requests.go

### PR DESCRIPTION
**Pull request by kayrus**
Fixes #2647

### CODE REVIEW

File: acceptance/clients/conditions.go [hunk 1]
```diff
@@ -147,7 +147,7 @@ func IsReleasesAbove(t *testing.T, release string) bool {
 func IsReleasesBelow(t *testing.T, release string) bool {
 	current_branch := getReleaseFromEnv(t)
 
-	if current_branch != "master" && current_branch < release {
+	if current_branch != "master" || current_branch < release {
```
Comment (kayrus) [inline acceptance/clients/conditions.go:150]: This fix allows to run tests on master branch only while using `clients.SkipReleasesBelow(t, "master")`

**Comment by kayrus** [comment]
@mandre @pierreprinetti ready for review

File: openstack/sharedfilesystems/v2/sharetransfers/results.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+type Transfer struct {
+	ID                   string              `json:"id"`
+	Accepted             bool                `json:"accepted"`
+	AuthKey              string              `json:"auth_key"`
+	Name                 string              `json:"name"`
+	SourceProjectID      string              `json:"source_project_id"`
+	DestinationProjectID string              `json:"destination_project_id"`
+	ResourceID           string              `json:"resource_id"`
+	ResourceType         string              `json:"resource_type"`
+	CreatedAt            time.Time           `json:"-"`
```
Comment (mandre) [inline openstack/sharedfilesystems/v2/sharetransfers/results.go:27]: Perhaps we could also add [`expires_at`](https://docs.openstack.org/api-ref/shared-file-system/?expanded=show-share-transfer-detail-detail#id774), WDYT?

File: openstack/sharedfilesystems/v2/sharetransfers/results.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+	Links                []map[string]string `json:"links"`
+}
+
+// UnmarshalJSON is our unmarshalling helper.
+func (r *Transfer) UnmarshalJSON(b []byte) error {
+	type tmp Transfer
+	var s struct {
+		tmp
+		CreatedAt gophercloud.JSONRFC3339MilliNoZ `json:"created_at"`
```
Comment (mandre) [inline openstack/sharedfilesystems/v2/sharetransfers/results.go:36]: Which makes me think, we still have https://github.com/gophercloud/gophercloud/issues/2507 :sweat_smile:

**Comment by kayrus** [comment]
You're right in all comments :) Thanks. As for the UTC offset, for some reason I cannot access debug logs for intermediate push :\ I don't have this API locally, therefore I rely on API reference and CI tests :)

**Comment by kayrus** [comment]
According to debug logs the time format is fine:

```
2023/06/21 07:48:00 [DEBUG] OpenStack Response Body: {
  "transfer": {
    "accepted": false,
    "auth_key": "4fbbb3ba15e6549e",
    "created_at": "2023-06-21T07:48:00.589548",
    "destination_project_id": null,
    "expires_at": "2023-06-21T07:53:00.587670",
    "id": "787da824-1b89-46f5-9ec7-e0b5afa2f28c",
    "links": [
      {
        "href": "http://10.1.120.0/share/v2/share-transfer/787da824-1b89-46f5-9ec7-e0b5afa2f28c",
        "rel": "self"
      },
      {
        "href": "http://10.1.120.0/share/share-transfer/787da824-1b89-46f5-9ec7-e0b5afa2f28c",
        "rel": "bookmark"
      }
    ],
    "name": "123",
    "resource_id": "dbae4463-25f3-42c3-9b18-46bc29ff733d",
    "resource_type": "share",    
    "source_project_id": "98df2063d7a8467eadc137dd4c5f4ca7"
  }
}
```

**Comment by mandre** [comment]
Yep, because our devstack returns the date without the UTC offset, which is why we're able to parse it. The `±hh:mm` part is optional in the [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) format. It would break if manila included it in it's responses.
It's not a big deal, as far as I know nobody reported issues, so that's probably not common at all in the wild, or perhaps just a documentation problem.

**Review by mandre — APPROVED** [review]

**Comment by kayrus** [comment]
@mandre BTW, does it make sense to turn on the `OS_DEBUG=true` for all tests?

**Comment by mandre** [comment]
> @mandre BTW, does it make sense to turn on the `OS_DEBUG=true` for all tests?

I'm afraid this is going to generate too much logs. The errors are generally obvious when we debug a CI job. We can always manually add it when needed, like you did.

**Pull request merged**